### DISCUSSION TITLE
Imitating darken/lighten less functions

### DISCUSSION THREAD

**Discussion by cmavelis** [system-ui/theme-ui#1599] [Q&A] [answered]
Maybe I'm overthinking it, but I was trying to decide how to mimic functionality of the color operations from less, for example: http://lesscss.org/functions/#color-operations-lighten

My goal was to add as little css as possible to my component library, so I've been writing the basic button states like `:hover` as part of the theme file.  If I want to simply darken the color in less, I use the darken function on the color variable.  

If I were to define a `darken` function in my theme.js, how could I apply it programmatically to the theme.js color variable? `darken('colors.primary')` for example, or do I need to define more colors to access them?

Edit: This looks like it would work if I were to redefine components using `sx` https://theme-ui.com/sx-prop#functional-values
However, I was hoping to do it via the theme.js config.  If that's not intended use for writing a component library, that's something I've been wondering as well!  https://github.com/tabetalt/kit Seems to define most things via config, and they've been my model.

Love the project btw, working on a component library prototype for my company as part of a new design system. I liked the look of rebass, saw the lack of activity, then found this!

**Comment by lachlanjc** [comment, marked as answer]
You're in luck! We've got a package just for that, with nearly exactly that API: https://theme-ui.com/packages/color

**Reply by cmavelis** [reply]
Ok so this implies I should add my hover states, etc. using the `sx` prop.  I haven't dug into how `theme.js` is parsed, but I imagine it would be possible to do this manipulation without redefining Button, for instance

**Reply by lachlanjc** [reply]
There's nothing to redefine—there's a certain set of styles built into the component itself, but then yeah, you can define theme-wide styles for components with variants: https://theme-ui.com/guides/variants/

**Reply by cmavelis** [reply]
variants are part of the `theme.js` config
color package is for use in the `sx` prop

What I'm asking is if I can use `darken()` in the theme.js file (for a globally defined hover style). My assumption is no, but your answer implies yes.
(if not, I'm pretty sure I'd need to define my own Button using `sx` for the darken fcn and export that as part of my component library)

**Reply by lachlanjc** [reply]
I've never tried it personally, but I'm pretty sure you can use the darken() etc directly in your theme. The color package has no special integration with the sx prop afaik—it's just a set of functions you can use that use the Theme UI theme context for your source color palette.

**Reply by cmavelis** [reply]
Ah ok! It does work, just needs to have rawColors defined.  That may be due to me using 0.4.0 since I was having issues with Storybook integration otherwise.

Thanks for all the answers!

**Reply by jjhamshaw** [reply]
@cmavelis I'm looking to do a similar thing - could you possibly share your solution? How do you use the rawColors?
I'm looking to do something like this in my theme config:
```
{
...
colors: {
    "background-1": darken("background", 0.1),
    "background-2": darken("background", 0.2),
    "background-3": darken("background", 0.3),
    modes: {
      dark: {
        "background-1": lighten("background", 0.1),
        "background-2": lighten("background", 0.2),
        "background-3": lighten("background", 0.3),
      },
    },
  },
}
```

**Reply by hasparus** [reply]
@jjhamshaw `lighten(name, amount)` returns a function which can't be used inside of `theme.colors` (i.e. colors can't be recursive), but this is perfectly doable if you move your base colors to a new variable.

```js

const baseColors = {
  background: '#ffffff'
}

const allColors = {
  'background-1': darken("background", 0.1)({ colors: baseColors })
}

```

I'm usually doing something like this when I need derived colors:

```ts
import { alpha, mix } from "@theme-ui/color";
import { map } from "fp-ts/lib/Record";

const light = {
  text: "#020202",
  gray: "#2B2C28",
  background: "#fff",
  // redacted for brevity
};

type BaseColorMode = typeof light;

const dark: BaseColorMode = {
  // redacted for brevity
};

const soft: BaseColorMode = {
  // redacted for brevity
};

const deriveAdditionalColors = (colors: BaseColorMode) => {
  return {
    ...colors,
    text092: alpha("text", 0.92)({ colors }),
    mutedPrimary09: mix("muted", "primary", 0.9)({ colors }),
  };
};

export type ColorMode = ReturnType<typeof deriveAdditionalColors>;

export const colorModes = map(deriveAdditionalColors)({
  light,
  dark,
  soft,
});
```

---

However, what you initially wrote is something we could possibly support in the future if more people want it. Would you like to write a feature request for it?

**Reply by cmavelis** [reply]
For now, @jjhamshaw, I've just copy-pasted my `colors` object to a `rawColors` object, but I'm not defining any other keyword colors like you are.  I'm going to do that programmatically, but your use case seems different.

@hasparus obviously has a good solution, but **I think** yours will work if you just add "background" as a key in rawColors:
```
{
...
rawColors: { background: #HEXHEX },
colors: {
    ...
  },
}
```

**Reply by hasparus** [reply]
@fcisio Do you remember when I mentioned we should split the theme provided by the user from the theme received from context on a type level? I said I'll _"get to it"... "some day"..._. Folks here started using `rawColors` as public API :cry: That's what I get for being lazy :D

---

> but I think yours will work if you just add "background" as a key in rawColors

It's not that I want your code not to work, guys, but it wasn't intended to work :D I'll be quite confused and a bit sad if it does work.

**Reply by fcisio** [reply]
😅  Yeah I was confused by this thread

**Reply by fcisio** [reply]
But for the general question on here, I have a solution that works pretty well!
`@theme-ui/color` uses `polished.js` behind the scene and convert theme value inside the functions.

But nothing is stopping us from using polished directly. Coupled with the Object color definition and the `__default` key, we have something pretty clean.

```js
import { darken, lighten } from 'polished'

const colors = {
  light: { __default: "#fff", 1: darken('#fff', 0.15), 2: darken('#fff', 0.35), 3: darken('#fff', 0.5) },
  dark: { __default: "#000", 1: lighten('#000', 0.15), 2: lighten('#000', 0.35), 3: lighten('#000', 0.5) },
}

// theme
colors: {
  background: colors.light,
  color: colors.dark.__default,
}
```

Then you can create builders to ease the process. Mine look a little something like that:

```js
makeSet(['light', '#fff'], darken, ['0.15', '0.35', '0.5'])
// outputs: { __default: "#fff", 1: darken('#fff', 0.15) , 2: darken('#fff', 0.35), 3: darken('#fff', 0.5) }
```
 
or in bulk with common values:

```js
makeSets(
  {
    light: [darken, '#ffffff'],
    dark: [lighten, '#000000'],
    yellow: [tint, '#FFBD2B'],
    blue: [shade, '#0759F9'],
  },
  [0.95, 0.75, 0.45, 0.15]
)
```

**Reply by cmavelis** [reply]
@hasparus I only do this because I couldn't get Storybook to work with `darken()` without having rawColors.  I vaguely remember getting an error message that said "use rawColors instead"

As for my suggestion, it doesn't work, so you can rest easy ;)

**Reply by fcisio** [reply]
Hi @cmavelis 
We recently introduced `rawColors` and it is a bit confusing. That's something we need to rethink.

Basically, in your theme, you should only have the `colors` key.
Then Theme UI will generate both `colors` and `rawColors`. In other words `rawColors` is strictly reserve to theme-ui and should not be defined by the user.

The error you are getting "use rawColors instead", is because you'll only find CSS Custom Properties in `colors` object (which you must be using in Storybook).
To fix this you can use the `rawColors` object since this once will always contain the actual color values. 

You can see more about this [here](https://dev.theme-ui.com/color-modes#colors-object)

**Reply by cmavelis** [reply]
To make sure I understand, can you show exactly what you mean by:
> To fix this you can use the rawColors object 

I have this pattern in my theme:
```
  colors: {
    text: '#161616',
    ...
  },
  rawColors: { //exact same as colors
    text: '#161616',
    ...
  }
```

**Reply by fcisio** [reply]
Sure:

```js
// You only define colors
colors: {
  text: '#161616',
  ...
},
// You don't want to define this
// rawColors: { //exact same as colors
//   text: '#161616',
//   ...
// }
```

The change need to happen where you use the colors:

```js
sx={(theme) => {
  bg: theme.colors.background, // = var(--theme-ui-colors-background)
  bg: theme.rawColors.background, // = #161616
}}
```

The error most likely surface if you are using `useThemeUI()`

```js
// look for
const { colors } = useThemeUI().theme

// if you then want the HEX value, change the code to this
const { rawColors: colors } = useThemeUI().theme
```

**Reply by cmavelis** [reply]
ok so is my problem coming from this usage pattern in my theme.js file?
```
...
buttons: {
    primary: {
      color: 'white',
      bg: darken('primary', 0.1),
      }
    }
    ...
```
maybe this is worth a new discussion, or sending me to somewhere this is already answered, but I'm simply trying to use my theme with this darken function in it in Storybook. I get the following error if I remove `rawColors`:
> Uncaught Error: Couldn't parse the color string. Please provide the color as a string in hex, rgb, rgba, hsl or hsla notation.

**Reply by hasparus** [reply]
Do you also get the same error in your app or does it happen only in Storybook?

Storybook also uses Emotion's context, and there are few possible mismatch bugs we still have to fix. (Having both Emotion v10 and Emotion v11 might be problematic.)

**Reply by cmavelis** [reply]
only storybook, and for context here are my versions:
`"theme-ui": "0.4.0-rc.1"`
Storybook 6.1.21 

and usage within `.storybook/preview.js`:
```
import React from 'react';

import { ThemeProvider, Container } from 'theme-ui';
import theme from '../../theme';

export const decorators = [
  (story) => (
  <ThemeProvider theme={theme}>
    <Container p={4}>{story()}</Container>
  </ThemeProvider>
),
];
```

**Reply by fcisio** [reply]
You can start by upgrading to the latest stable version.
It would be hard for anyone to debug on `0.4.0-rc.1` which is severely outdated by now.

> I don't think `0.4.0-rc.1` even has the `rawColors` object in the context.

As for @hasparus point above, this is indeed most likely an Emotion version mismatch

Here's a snippet with resolution on Emotion. Set this in your `main.js` file and it should fix any mismatch

```js
module.exports = ({ config }) => {
  config.resolve.alias = {
    ...config.resolve.alias,
    "@emotion/core": toPath("node_modules/@emotion/react"),
    "@emotion/styled": toPath("node_modules/@emotion/styled"),
    "emotion-theming": toPath("node_modules/@emotion/react"),
  }

  return config
}
```

**Reply by hasparus** [reply]
+1 to the idea about updating. I mean, any stable version should work, but old `alpha` or `rc` might be problematic.

**Reply by cmavelis** [reply]
ok we're updated and working without `rawColors`! 🎉 Here's how to implement that code snippet you sent in current versions of Storybook (alongside the rest of my config):
`.storybook/main.js`:
```js
const path = require('path')

module.exports = {
  "webpackFinal": (config) => {
    config.resolve.alias = {
      ...config.resolve.alias,
      "@emotion/core": path.resolve("node_modules/@emotion/react"),
      "@emotion/styled": path.resolve("node_modules/@emotion/styled"),
      "emotion-theming": path.resolve("node_modules/@emotion/react"),
    }
    return config
  },
  "stories": [
    "../src/**/*.stories.mdx",
    "../src/**/*.stories.@(js|jsx|ts|tsx)"
  ],
  "addons": [
    "@storybook/addon-links",
    "@storybook/addon-essentials"
  ]
}
```
